File: app/src/main/java/com/better/alarm/statemachine/ComplexTransition.java
package com.better.alarm.statemachine;

public abstract class ComplexTransition extends State {
    abstract public void performComplexTransition();

    @Override
    public final void enter() {
        performComplexTransition();
    }

    @Override
    public final void resume() {
        //empty
    }

    @Override
    public final boolean processMessage(Message msg) {
        throw new RuntimeException("performComplexTransition() must transit immediately");
    }

    @Override
    public final void exit() {
        // nothing to do
    }
}

File: app/src/main/java/com/better/alarm/domain/statemachine/ComplexTransition.kt
package com.better.alarm.domain.statemachine

abstract class ComplexTransition<T> : State<T>() {
  abstract fun performComplexTransition()

  override fun enter(reason: T?) {
    performComplexTransition()
  }

  override fun onEvent(event: T): Boolean {
    throw RuntimeException("performComplexTransition() must transit immediately")
  }

  override fun exit(reason: T?) {
    // nothing to do
  }
}
